libs/vkd3d/command.c

```diff
@@ -11879,6 +12199,88 @@ static void d3d12_command_queue_signal(struct d3d12_command_queue *command_queue
     /* We should probably trigger DEVICE_REMOVED if we hit any errors in the submission thread. */
 }
 
+static void d3d12_command_queue_wait_shared(struct d3d12_command_queue *command_queue,
+        struct d3d12_shared_fence *fence, UINT64 value)
+{
+    assert(fence->timeline_semaphore);
+    vkd3d_queue_add_wait(command_queue->vkd3d_queue, &fence->ID3D12Fence_iface, fence->timeline_semaphore, value);
+}
+
+static void d3d12_command_queue_signal_shared(struct d3d12_command_queue *command_queue,
+        struct d3d12_shared_fence *fence, UINT64 value)
+{
+    VkTimelineSemaphoreSubmitInfoKHR timeline_submit_info;
+    const struct vkd3d_vk_device_procs *vk_procs;
+    VkSemaphore signal_semaphores[2];
+    struct vkd3d_queue *vkd3d_queue;
+    struct d3d12_device *device;
+    uint64_t signal_values[2];
+    VkSubmitInfo submit_info;
+    VkQueue vk_queue;
+    VkResult vr;
+
+    device = command_queue->device;
+    vk_procs = &device->vk_procs;
+    vkd3d_queue = command_queue->vkd3d_queue;
+
+    TRACE("queue %p, fence %p, value %#"PRIx64".\n", command_queue, fence, value);
+
+    if (!(vk_queue = vkd3d_queue_acquire(vkd3d_queue)))
+    {
+        ERR("Failed to acquire queue %p.\n", vkd3d_queue);
+        return;
+    }
+
+    timeline_submit_info.sType = VK_STRUCTURE_TYPE_TIMELINE_SEMAPHORE_SUBMIT_INFO_KHR;
+    timeline_submit_info.pNext = NULL;
+
+    submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+    submit_info.pNext = &timeline_submit_info;
+    submit_info.commandBufferCount = 0;
+    submit_info.pCommandBuffers = NULL;
+
+    /* We need to use the submission timeline as we don't have a physical value for shared fences */
+    vkd3d_queue->submission_timeline_count++;
+
+    signal_semaphores[0] = fence->timeline_semaphore;
+    signal_semaphores[1] = vkd3d_queue->submission_timeline;
+
+    submit_info.pSignalSemaphores = signal_semaphores;
+    submit_info.signalSemaphoreCount = 2;
+
+    signal_values[0] = value;
+    signal_values[1] = vkd3d_queue->submission_timeline_count;
+
+    timeline_submit_info.pSignalSemaphoreValues = signal_values;
+    timeline_submit_info.signalSemaphoreValueCount = 2;
```

Comment:
If we want submissions timeline to  come after the signal of shared fence, we need two submission batches to guarantee the order, otherwise the two semaphores here can signal out of order leading to a subtle race if we free the shared fence based on submission timeline signal.